### PR TITLE
[AssetMapper] Add cached asset factory

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | None
| License       | MIT
| Doc PR        | Still TODO

Hi!

This is built on top of #50264. I CAN separate them if needed - sorry for the noise.

tl;dr When using asset mapper, to load the page, we need to build the importmap & resolve a few paths, like `{{ asset('styles/app.css') }}`. Because asset mapper loads the contents and performs some regex on CSS and JS files, this can slow down the page in dev (in prod, `asset-map:compile` removes all overhead).

This PR fixes that (I can see a big improvement in my test app). We "cache" the `MappedAsset` objects via the config cache system so that if an underlying file changes (e.g. `assets/styles/app.css`), we only need to rebuild that ONE `MappedAsset` when the page is loading. 

Cheers!

### CODE REVIEW

File: src/Symfony/Component/AssetMapper/AssetMapperCompiler.php [hunk 1]
```diff
@@ -25,20 +26,25 @@ class AssetMapperCompiler
     /**
      * @param iterable<AssetCompilerInterface> $assetCompilers
      */
-    public function __construct(private iterable $assetCompilers)
+    public function __construct(private iterable $assetCompilers, private ContainerInterface $container)
```
Comment (weaverryan) [inline src/Symfony/Component/AssetMapper/AssetMapperCompiler.php]: This is the ugliest part of this PR, and it's actually a small fix so that decorating the AssetMapper will work properly.

* It's super convenient for compilers to have access to the AssetMapper
* But, we get a circular reference: AssetMapper -> MappedAssetFactory -> AssetMapperCompiler -> AssetMapper

Is there a more elegant way to make *just* this one dependency lazy?

File: src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+use Symfony\Component\AssetMapper\MappedAsset;
+use Symfony\Component\Config\ConfigCache;
+use Symfony\Component\Config\Resource\FileResource;
+
+class CachedMappedAssetFactory implements MappedAssetFactoryInterface
+{
+    public function __construct(
+        private MappedAssetFactoryInterface $innerFactory,
+        private string $cacheDir,
```
Comment (stof) [inline src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php]: should they be readonly, like in other classes in this component ?

File: src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            return null;
+        }
+
+        $resources = [new FileResource($mappedAsset->getSourcePath())];
+        foreach ($mappedAsset->getDependencies() as $dependency) {
+            if (!$dependency->contentDependency) {
+                continue;
+            }
+
+            $resources[] = new FileResource($dependency->asset->getSourcePath());
```
Comment (stof) [inline src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php]: what about recursive dependencies ? Are they handled properly ?

File: src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        }
+        $configCache->write(serialize($mappedAsset), $resources);
+
+        return $mappedAsset;
+    }
+
+    private function getCacheFilePath(string $logicalPath, string $sourcePath): string
+    {
+        return $this->cacheDir . '/mapped_assets_cache_' . md5($logicalPath.':'.$sourcePath) . '.php';
```
Comment (stof) [inline src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php]: why using md5 here and not xxh128 which is faster ?

Comment (stof) [inline src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php]: Instead of a `mapped_assets_cache_` prefix on files being directly in the cache folder, I think this cache should be in a dedicated subfolder, as done for most caches in FrameworkBundle (which would then be the responsibility of the bundle when configuring the cache directory)

File: src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        private MappedAssetFactoryInterface $innerFactory,
+        private string $cacheDir,
+    )
+    {
+    }
+
+    public function createMappedAsset(string $logicalPath, string $sourcePath): ?MappedAsset
+    {
+        $cachePath = $this->getCacheFilePath($logicalPath, $sourcePath);
+        $configCache = new ConfigCache($cachePath, true);
```
Comment (stof) [inline src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php]: should we always enable debug mode ?

File: src/Symfony/Component/AssetMapper/AssetDependency.php [hunk 6]
```diff
@@ -19,11 +19,20 @@
 final class AssetDependency
 {
     /**
-     * @param bool $isLazy whether this dependency is immediately needed
+     * @param bool $isLazy Whether the dependent asset will need to be loaded eagerly
+     *                     by the parent asset (e.g. a CSS file that imports another
+     *                     CSS file) or if it will be loaded lazily (e.g. an async
+     *                     JavaScript import).
+     * @param bool $contentDependency Whether the parent asset's content depends
```
Comment (stof) [inline src/Symfony/Component/AssetMapper/AssetDependency.php]: I don't see any place where you configure a dependency as a content dependency. Any asset compiler retrieving the public path of a dependency needs to do it.

File: src/Symfony/Component/AssetMapper/MappedAsset.php [hunk 7]
```diff
@@ -120,9 +120,9 @@ public function setContent(string $content): void
         $this->content = $content;
     }
 
-    public function addDependency(self $asset, bool $isLazy = false): void
+    public function addDependency(self $asset, bool $isLazy = false, bool $contentDependency = false): void
```
Comment (stof) [inline src/Symfony/Component/AssetMapper/MappedAsset.php]: this method needs to have phpdoc describing those 2 arguments, as they are the ones that implementors of asset compilers will see in their IDE when using the code, not the AssetDependency constructor (alternatively, we might change this to `->addDependency(AssetDependency $dependency)` to avoid having to duplicate the signature)

Comment (stof) [inline src/Symfony/Component/AssetMapper/MappedAsset.php]: shouldn't `$contentDependency` default to `true` to be safer ? Marking dependencies as affecting the content when they are not would lead to useless cache invalidation (which only makes thing slower) while the opposite leads to returning the wrong result.

File: src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+        $resources = [new FileResource($mappedAsset->getSourcePath())];
+        foreach ($mappedAsset->getDependencies() as $dependency) {
+            if (!$dependency->contentDependency) {
+                continue;
+            }
+
+            $resources[] = new FileResource($dependency->asset->getSourcePath());
+        }
+        $configCache->write(serialize($mappedAsset), $resources);
```
Comment (stof) [inline src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php]: Given that the MappedAsset object contains AssetDependency objects that contain other MappedAsset objects, what happens when you serialize and unserialize those ?

Also, what is the impact of unserializing a MappedAsset with dependencies that will have MappedAsset objects not known by the AssetMapper internal cache (while potentially referring to the same assets) and also different from MappedAsset created for dependencies of other cached MappedAsset ?

File: src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php (hunk 3 above)
Comment (weaverryan) [reply src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php]: they're not! Thx - fixing now.

File: src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php (hunk 4 above)
Comment (weaverryan) [reply src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php]: I'm using `param('kernel.cache_dir').'/asset_mapper',` in the `asset_mapper.php` service config. It's not configurable like other systems... but does it need to be?

You're right about this `mapped_assets_cache_` prefix - that's redundant - removing that.

File: src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php (hunk 5 above)
Comment (weaverryan) [reply src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php]: It's a good question. In practice, this system will never be used in production, because you'll have dumped your `manifest.json` and `importmap.json` files, so only those will be used. 

But, might as well. I've updated it to use `%kernel.debug%`.

File: src/Symfony/Component/AssetMapper/AssetDependency.php (hunk 6 above)
Comment (weaverryan) [reply src/Symfony/Component/AssetMapper/AssetDependency.php]: ty - great catch

File: src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php (hunk 4 above)
Comment (stof) [reply src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php]: it does not need to be configurable (others are not either). When reviewing this code, I did not check the service config to see whether the prefix was just redundant or if it was there to put files directly in the main cache folder.

File: src/Symfony/Component/AssetMapper/AssetMapperCompiler.php (hunk 1 above)
Comment (stof) [reply src/Symfony/Component/AssetMapper/AssetMapperCompiler.php]: You could inject a lazy proxy here, using the same kind of trick than used when using `#[Autowire(lazy: AssetMapperInterface::class)]` for autowiring. But this would make the component much harder to use standalone if the circular dependency is between constructors (as the _only_ way to instantiate the object graph would be through a lazy proxy)

File: src/Symfony/Component/AssetMapper/MappedAsset.php (hunk 7 above)
Comment (weaverryan) [reply src/Symfony/Component/AssetMapper/MappedAsset.php]: Good idea just to use `->addDependency(AssetDependency $dependency)`

File: src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php (hunk 8 above)
Comment (weaverryan) [reply src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php]: > Given that the MappedAsset object contains AssetDependency objects that contain other MappedAsset objects, what happens when you serialize and unserialize those ?

It seems to work fine, in practice. Both `MappedAsset` and `AssetDependency` are objects that hold scalars. For example, here is a dump of one situation (dumping the result of the `unserialize` in the cached factory) in a test project:

Screenshot 2023-05-10 at 1 32 25 PM

What concerns do you have?

> Also, what is the impact of unserializing a MappedAsset with dependencies that will have MappedAsset objects not known by the AssetMapper internal cache (while potentially referring to the same assets) and also different from MappedAsset created for dependencies of other cached MappedAsset ?

I think I understand what you're referring to. So, for example:

* `asset1.css` has a dependency on `asset2.css`. When something asks for the `asset1.css` asset, its cache is fresh, so we get back `MappedAsset` instances from the cache for both `asset1.css` and `asset2.css`.
* Meanwhile, `asset3.css` also has a dependency on `asset2.css`. When something asks for the `asset3.css` asset, its cache is NOT fresh, so we get back FRESH `MappedAsset `instances for both `asset3.css` and `asset2.css`.

So, at this moment, there are two distinct `asset2.css` `MappedAsset` objects floating around the system. Their data is equal, but the objects are not identical. Other than being wasteful, I don't think it's a problem. `MappedAsset` objects are "read-only" informational objects. Well, except for `MappedAsset::addDependency()`. But, in practice, once a `MappedAsset` is returned from `AssetMapper`, it is intended to be read-only: its data is entirely set *during* the creation process (we could even "enforce" this by adding some flag to "freeze" the `MappedAsset` and thus reject calling `addDependency()`, but that might be overkill).

However, if we're concerned about this, we *could*, "refresh" the dependencies after fetching a cached `MappedAsset` - e.g. loop over `$mappedAsset->getDependencies()`... grab the `->logicalName` + `->getSourcePath()` of each, call `$this->createMappedAsset($logicalName, $sourcePath)` on each, then "replace" the `MappedAsset` on each dependency.

tl;dr I don't see any real problem, but I could add extra code to be absolutely sure.

File: src/Symfony/Component/AssetMapper/AssetMapperCompiler.php (hunk 1 above)
Comment (weaverryan) [reply src/Symfony/Component/AssetMapper/AssetMapperCompiler.php]: > But this would make the component much harder to use standalone

Thank you - that makes me want to keep this as it is :)

**Comment by fabpot** [comment]
Should be rebased :)

**Comment by weaverryan** [comment]
Rebased, reviewed again (by me) and tested again in my local app. Looks good 👍

File: src/Symfony/Component/AssetMapper/AssetMapperCompiler.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/AssetMapper/AssetMapperCompiler.php]: what about using a closure here, that'd return the mapper? that's a service_closure inthe DI world

File: src/Symfony/Component/AssetMapper/Factory/MappedAssetFactory.php [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    private function getPublicPath(MappedAsset $asset): ?string
+    {
+        [$digest, $isPredigested] = $this->getDigest($asset);
+
+        if ($isPredigested) {
+            return $this->assetsPathResolver->resolvePublicPath($asset->logicalPath);
+        }
+
+        $digestedPath = preg_replace_callback('/\.(\w+)$/', function ($matches) use ($digest) {
+            return "-{$digest}{$matches[0]}";
```
Comment (nicolas-grekas) [inline src/Symfony/Component/AssetMapper/Factory/MappedAssetFactory.php]: short closure FTW

File: src/Symfony/Component/AssetMapper/Tests/Compiler/JavaScriptImportPathCompilerTest.php [hunk 10]
```diff
@@ -34,6 +34,9 @@ public function testCompile(string $sourceLogicalName, string $input, array $exp
             $actualDependencies[$dependency->asset->logicalPath] = $dependency->isLazy;
         }
         $this->assertEquals($expectedDependencies, $actualDependencies);
+        if (\count($expectedDependencies) > 0) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/AssetMapper/Tests/Compiler/JavaScriptImportPathCompilerTest.php]: we never "count" on array: 
```suggestion
        if ($expectedDependencies) {
```

File: src/Symfony/Component/AssetMapper/AssetMapper.php [hunk 11]
```diff
@@ -23,50 +24,24 @@
...
-            $asset->setContent($this->calculateContent($asset));
-
-            array_pop($this->assetsBeingCreated);
+        $filePath = $this->mapperRepository->find($logicalPath);
+        if (null === $filePath) {
+            return null;
         }
 
-        return $this->assetsCache[$logicalPath];
+        return $this->mappedAssetFactory->createMappedAsset($logicalPath, $filePath);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/AssetMapper/AssetMapper.php]: can't this get `$this` as argument to remove the need for the container/service-closure on the compiler?

Comment (stof) [reply src/Symfony/Component/AssetMapper/AssetMapper.php]: passing `$this` will make this incompatible with decorating the AssetMapper. as it would not pass the decorator.

Not sure it is needed though. This feature relies on decorating the asset factory, not the AssetMapper.

Comment (weaverryan) [reply src/Symfony/Component/AssetMapper/AssetMapper.php]: Stof is right. If someone decorated the main `AssetMapper`, then the compilers would get the core, non-decorated one. It's... a bit unfortunate, I admit.

File: src/Symfony/Component/AssetMapper/AssetMapperCompiler.php (hunk 1 above)
Comment (fabpot) [reply src/Symfony/Component/AssetMapper/AssetMapperCompiler.php]: I don't think we can keep this as is TBH.

Comment (weaverryan) [reply src/Symfony/Component/AssetMapper/AssetMapperCompiler.php]: I'll fix this up shortly!

**Comment by weaverryan** [comment]
This should be happy now - `service_closure` added - much, much nicer.

File: src/Symfony/Component/AssetMapper/AssetMapperCompiler.php [hunk 12]
```diff
@@ -22,23 +22,34 @@
...
+            $content = $compiler->compile($content, $mappedAsset, $this->getAssetMapper());
         }
 
         return $content;
     }
+
+    private function getAssetMapper(): AssetMapperInterface
+    {
+        if (!isset($this->assetMapper)) {
+            $this->assetMapper = $this->assetMapperFactory->__invoke();
```
Comment (fabpot) [inline src/Symfony/Component/AssetMapper/AssetMapperCompiler.php]: Would that work?

```suggestion
            $this->assetMapper = ($this->assetMapperFactory)();
```

File: src/Symfony/Component/AssetMapper/AssetMapperCompiler.php [hunk 13]
```diff
@@ -22,23 +22,34 @@
  */
 class AssetMapperCompiler
 {
+    private AssetMapperInterface $assetMapper;
+
     /**
      * @param iterable<AssetCompilerInterface> $assetCompilers
```
Comment (stof) [inline src/Symfony/Component/AssetMapper/AssetMapperCompiler.php]: I suggest adding `@param \Closure(): AssetMapperInterface $assetMapperFactory` to document the expected signature of the closure

File: src/Symfony/Component/AssetMapper/Compiler/CssAssetUrlCompiler.php [hunk 14]
```diff
@@ -47,7 +48,7 @@ public function compile(string $content, MappedAsset $asset, AssetMapperInterfac
                 return $matches[0];
             }
 
-            $asset->addDependency($dependentAsset);
+            $asset->addDependency(new AssetDependency($dependentAsset, isContentDependency: true));
```
Comment (stof) [inline src/Symfony/Component/AssetMapper/Compiler/CssAssetUrlCompiler.php]: This is the default for `isContentDependency` here.

File: src/Symfony/Component/AssetMapper/AssetMapperCompiler.php [hunk 15]
```diff
@@ -22,23 +22,35 @@
...
-    public function compile(string $content, MappedAsset $mappedAsset, AssetMapperInterface $assetMapper): string
+    public function compile(string $content, MappedAsset $mappedAsset): string
     {
         foreach ($this->assetCompilers as $compiler) {
             if (!$compiler->supports($mappedAsset)) {
                 continue;
             }
 
-            $content = $compiler->compile($content, $mappedAsset, $assetMapper);
+            $content = $compiler->compile($content, $mappedAsset, $this->getAssetMapper());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/AssetMapper/AssetMapperCompiler.php]:
```suggestion
            $content = $compiler->compile($content, $mappedAsset, $this->assetMapper ??= ($this->assetMapperFactory)());
```

File: src/Symfony/Component/AssetMapper/Factory/MappedAssetFactory.php [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+    /**
+     * Returns an array of "string digest" and "bool predigested".
+     *
+     * @return array{0: string, 1: bool}
+     */
+    private function getDigest(MappedAsset $asset): array
+    {
+        // check for a pre-digested file
+        if (1 === preg_match(self::PREDIGESTED_REGEX, $asset->getLogicalPath(), $matches)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/AssetMapper/Factory/MappedAssetFactory.php]:
```suggestion
        if (self::PREDIGESTED_REGEX, $asset->getLogicalPath(), $matches)) {
```

**Review by nicolas-grekas — APPROVED** [review]

**Comment by weaverryan** [comment]
Updated - thanks for the suggestions :)

File: src/Symfony/Component/AssetMapper/Factory/MappedAssetFactory.php (hunk 16 above)
Comment (stof) [reply src/Symfony/Component/AssetMapper/Factory/MappedAssetFactory.php]: the suggestion removes too many things

Comment (weaverryan) [reply src/Symfony/Component/AssetMapper/Factory/MappedAssetFactory.php]: you're correct - but I got it right when I applied it:

```php
if (preg_match(self::PREDIGESTED_REGEX, $asset->getLogicalPath(), $matches))
```

**Comment by fabpot** [comment]
Thank you @weaverryan.
